chevah/python-package
# [#3855] Custom Python versioning on Linux/Unix.

Scope
=====

The `python` binary in `python-package` should advertise a custom version through `sys.version`.
The custom version should be based on the current git revision.
The python archive name should include this custom version, eg. `python-2.7.10.c3dea23-linux-x86.tar.gz`.

Changes
=======
Added a `sed` command in python's `chevahbs` script to hijack an unused Mercurial-related string which would have always been `default` downstream of us. This is how the result looks like (the short form of the git revision is displayed before the build date):

```
Python 2.7.10 (ba67e06, Feb 13 2017, 14:08:38) 
[GCC 4.8.4] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> sys.version
>>> '2.7.10 (ba67e06, Feb 13 2017, 14:08:38) \n[GCC 4.8.4]'
```
Added a python test for checking the presence of the short form of the git revision in python's `sys.version` string. When failing, the check errors out as follows:
```
% ./chevah_build test
~/dumol/python-package/build ~/dumol/python-package
python 2.7.10 (18aebdb9, Feb 13 2017, 18:38:53) 
[GCC 4.8.4]
zlib 1.0
stdlib ssl OpenSSL 1.0.1f 6 Jan 2014
cryptography 1.3.1 - OpenSSL OpenSSL 1.0.1f 6 Jan 2014
pyopenssl 16.0.0 - OpenSSL OpenSSL 1.0.1f 6 Jan 2014
PyCrypto 2.6.1
ctypes 1.1.0
setproctitle 1.1.10
Python's version doesn't match git rev!
        Bin ver: 18aebdb
        Git rev: a896862
PWD : /home/chevah/dumol/python-package/build
Fail: ./python2.7-ubuntu1404-x64/bin/python test_python_binary_dist.py

```
Added the git rev string in the python package archive and stopped using time stamps.

**Drive-by changes**:
  * removed duplicate python tests
  * cleaned up the error code scheme even more.

How to try and test the changes
===============================

reviewers: @adiroiban

Please review the changes.
Build and test `python-package`, eg. https://chevah.com/buildbot/builders/python-package-gk-review/builds/29
Check `sys.version` for the `python` binary.
Check the file name for the newly created archive.
Check that the test for the git rev works (eg. by testing with a different python binary).
